lib/Core/AddressSpace.cpp

```diff
@@ -160,125 +160,118 @@ bool AddressSpace::resolveOne(ExecutionState &state,
   }
 }
 
-bool AddressSpace::resolve(ExecutionState &state,
-                           TimingSolver *solver, 
-                           ref<Expr> p, 
-                           ResolutionList &rl, 
-                           unsigned maxResolutions,
-                           double timeout) {
+int AddressSpace::checkPointerInObject(ExecutionState &state,
+                                       TimingSolver *solver, ref<Expr> p,
+                                       const ObjectPair &op, ResolutionList &rl,
+                                       unsigned maxResolutions) {
+  // XXX in the common case we can save one query if we ask
+  // mustBeTrue before mayBeTrue for the first result. easy
+  // to add I just want to have a nice symbolic test case first.
+  const MemoryObject *mo = op.first;
+  ref<Expr> inBounds = mo->getBoundsCheckPointer(p);
+  bool mayBeTrue;
+  if (!solver->mayBeTrue(state, inBounds, mayBeTrue)) {
+    return 1;
+  }
+
+  if (mayBeTrue) {
+    llvm::errs() << "Adding object to RL\n";
+    rl.push_back(op);
+
+    // fast path check
+    unsigned size = rl.size();
+    if (size == 1) {
+      bool mustBeTrue;
+      if (!solver->mustBeTrue(state, inBounds, mustBeTrue))
+        return 1;
+      if (mustBeTrue)
+        return 0;
+    }
+    else
+      if (size == maxResolutions)
+        return 1;
+  }
+
+  return 2;
+}
+
```

Comment:
IMHO this function would benefit from an `early exit` style. It would reduce the nesting and increase the readability. Also returning named values (define, enum) would be nice.

Reply:
I'm not sure I see more opportunities for early exit.   What do you have in mind?

Reply:
`if(!mayBeTrue) return 2;`
`if (size == maxResolutions && size != 1)  return 1;`

I think.

Reply:
untested:
```
  // 0: ok, 1: solver error or maxium reached, 2: multiple resolutions
 
 ...
  // make sure pointer is in bounds
  if (!solver->mayBeTrue(state, inBounds, mayBeTrue)) return 1;
  if (!mayBeTrue) return 2;
 
  // add object to resolution list
  rl.push_back(op);
  unsigned size = rl.size();

  // exit on multiple resolutions or reached maximum
  if (size != 1)
    return size == maxResolutions ? 1 : 2;

  // check if this object is the only solution
  bool mustBeTrue;
  if (!solver->mustBeTrue(state, inBounds, mustBeTrue)) return 1;
  return mustBeTrue ? 0 : 2;
```

Reply:
I find the flow to be more natural as it is, but I guess it's a matter of style.

Reply:
@251 I see you're a big fun of early exits :)   I actually find nested ifs to be often nicer, because I can quickly see by the level of nesting where we are in the program.   But again, it's a matter of preference.  And as you can tell, I'm a big fan of not imposing a particular style, although I realize there are advantages both ways.

Reply:
I consider nested `if`s an antipattern and find the code hard to read esp. with 2 space indentation level. I wouldn't even know how to add comments to all those branches...